test/cpp/jit/test_lite_interpreter.cpp

```diff
@@ -524,7 +524,7 @@ TEST(LiteInterpreterTest, SequentialModuleInfo) {
   }
 
   std::unordered_set<std::string> expected_result(
-      {"top(C).A0(A).forward", "top(C).B0(B).forward"});
+      {"top(C).A0(A).forward", "top(C).A0(A).forward.B0(B).forward"});
```

Comment:
Module C forward methods is `return self.A0.forward(self.B0.forward(x))`, so the expected result should be `top(C).A0(A).forward.B0(B).forward`

Reply:
Can you actually check and see if this is the output you get.
The reason I am saying this is the code `return self.A0.forward(self.B0.forward(x))` is equivalent of:
```
z = self.B0.forward(x)
return self.A0.forward(z)
```
So B0.forward gets called from top module and A0.forward as well. The changes you have would be valid if B0.forward was getting called from A0.forward.

Reply:
Thanks for sharing the example! I check the result and it is the output I got. Will investigate what goes wrong here.

Reply:
I *think* (and I could be wrong), this should just be 2 separate calls `"top(C).A0(A).forward"` and `"top(C).B0(B).forward"`
 since I don't think there is any code level nesting in here. I think the call-stack is meant to capture the lexical nesting and not the control flow nesting (I could be wrong).

Reply:
I debugged through the code. This line, https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/ir/ir.cpp#L1916 looks confusing to me. My understanding is that it tries to go through the inserted new nodes (new_node), and set callstack to new_node. However, value_map is not a node:node map. It's input:input and output:output map. Is it possible that it collects a new_node that is not from callee? For example, we I debugged through this example, there are two points I don't fully get: 1) `raw_callstack_ptr` is always NULL for the actual `add` node. 2) It first inlines B.forward. When it goes A.forward, in addition to checking new_node = `add` node from A, it also checks new_node = inserted `add` node from B, which does not look proper to me. @ZolotukhinM, could you provide more background to the callstack setup?

Reply:
@dhruvbird is correct, in @kimishpatel 's example the IR would have two separate nodes for the two forward calls and each of them would have its own callstack - they won't be (and should not be) combined.

Reply:
@iseeyuan, the code tries to append a new entry to callstacks of nodes of the graph we've inlined. Callstacks are implemented as a linked list, so we're essentially appending an entry at the head of the list:
https://github.com/pytorch/pytorch/blob/9ea7a6c7c55484b83b8cedb58158f429425fe950/torch/csrc/jit/ir/ir.cpp#L1929-L1934
or start a new list if the callstack was NULL:
https://github.com/pytorch/pytorch/blob/9ea7a6c7c55484b83b8cedb58158f429425fe950/torch/csrc/jit/ir/ir.cpp#L1936-L1938

In this example the IR after inlining would roughly look like:
```
graph(...):
  %t1 = aten::add(...) # callstack: [A0::forward]
  ...
  %t2 = aten::add(%t1, ...) # callstack: [B0::forward]
  ...
```

In order to get an example with a callstack of depth > 1, one would need to write something like this:
```
 class A(nn.Module):
   def __init__(self):
     super(A, self).__init__()

   def forward(self, x):
     return x + 1

 class B(nn.Module):
   def __init__(self):
     super(B, self).__init__()
     self.A_from_B = A()

   def forward(self, x):
     return self.A_from_B.forward(x) + 2

 class C(nn.Module):
   def __init__(self):
     super(C, self).__init__()
     self.B_from_C = B()

   def forward(self, x):
     return self.B_from_C.forward(x)
```

Reply:
@ZolotukhinM I followed Martin's logic and also dig through. In the latest commit, I append a `unodered_map<Node*, Node*> node_map` and use it to get the new node, instead of from `value_map`. With this change, the result is `"top(C).A0(A).forward", "top(C).B0(B).forward"`. The latest commit includes some hack and I will try polish it. Do you have any concern regarding using `node_map` instead of `value_map`?

Reply:
Chen, thanks for digging into this somewhat convoluted issue. Can you elaborate on why you need node_map? Is the node that you get from value_map is not right? From my understanding of your changes it should be.

Reply:
The node I got from value_map (`Node* new_node = kv.second->node()`) is different from the one I got from node_map(`Node* new_node = kv.second->node()`) when inlining C.forward()

```
before inline: graph(%self.1 : __torch__.C,
      %x.1 : Tensor):
  %2 : __torch__.A = prim::GetAttr[name="A0"](%self.1)
  %3 : __torch__.B = prim::GetAttr[name="B0"](%self.1)
  %4 : Tensor = prim::CallMethod[name="forward"](%3, %x.1) # <string>:3:29
  %5 : Tensor = prim::CallMethod[name="forward"](%2, %4) # <string>:3:13
  return (%5)

before inline: graph(%self : __torch__.B,
      %x.1 : Tensor):
  %2 : int = prim::Constant[value=1]()
  %3 : int = prim::Constant[value=2]() # <string>:3:17
  %4 : Tensor = aten::add(%x.1, %3, %2) # <string>:3:13
  return (%4)

after inline: graph(%self : __torch__.B,
      %x.1 : Tensor):
  %2 : int = prim::Constant[value=1]()
  %3 : int = prim::Constant[value=2]() # <string>:3:17
  %4 : Tensor = aten::add(%x.1, %3, %2) # <string>:3:13
  return (%4)

listing nodes from value map 
%self.1 : __torch__.C, %x.1 : Tensor = prim::Param()

%3 : __torch__.B = prim::GetAttr[name="B0"](%self.1)

listing nodes from node map 
%6 : int = prim::Constant[value=2]() # <string>:3:17

%7 : int = prim::Constant[value=1]()

%8 : Tensor = aten::add(%x.1, %6, %7) # <string>:3:13

finish listing. 
before inline: graph(%self : __torch__.A,
      %x.1 : Tensor):
  %2 : int = prim::Constant[value=1]() # <string>:3:17
  %3 : Tensor = aten::add(%x.1, %2, %2) # <string>:3:13
  return (%3)

after inline: graph(%self : __torch__.A,
      %x.1 : Tensor):
  %2 : int = prim::Constant[value=1]() # <string>:3:17
  %3 : Tensor = aten::add(%x.1, %2, %2) # <string>:3:13
  return (%3)

listing nodes from value map 
%8 : Tensor = aten::add(%x.1, %6, %7) # <string>:3:13

%2 : __torch__.A = prim::GetAttr[name="A0"](%self.1)

listing nodes from node map 
%9 : int = prim::Constant[value=1]() # <string>:3:17

%10 : Tensor = aten::add(%8, %9, %9) # <string>:3:13

finish listing. 
after inline: graph(%self.1 : __torch__.C,
      %x.1 : Tensor):
  %2 : __torch__.A = prim::GetAttr[name="A0"](%self.1)
  %3 : __torch__.B = prim::GetAttr[name="B0"](%self.1)
  %6 : int = prim::Constant[value=2]() # <string>:3:17
  %7 : int = prim::Constant[value=1]()
  %8 : Tensor = aten::add(%x.1, %6, %7) # <string>:3:13
  %9 : int = prim::Constant[value=1]() # <string>:3:17
  %10 : Tensor = aten::add(%8, %9, %9) # <string>:3:13
  return (%10)
```

Reply:
This definitely seems wrong to me. Can you add this check, `TORCH_CHECK(new_node->outputs()[i] == new_node, "Both nodes should be equal");`, here: https://github.com/pytorch/pytorch/pull/47416/commits/ce308cde64ce648984189610060ad828c18b63a7#diff-12d743f00439e8ccdc9575e788d00b60081dfbc8ce0a9e7b5184cb0d87843c2aR2030.
Reason why I am suggesting this is that if mapped value is pointing to some other node then this is a bug, that needs to be fixed.

Reply:
> @iseeyuan, the code tries to append a new entry to callstacks of nodes of the graph we've inlined. Callstacks are implemented as a linked list, so we're essentially appending an entry at the head of the list:
> https://github.com/pytorch/pytorch/blob/9ea7a6c7c55484b83b8cedb58158f429425fe950/torch/csrc/jit/ir/ir.cpp#L1929-L1934
> 
> 
> or start a new list if the callstack was NULL:
> https://github.com/pytorch/pytorch/blob/9ea7a6c7c55484b83b8cedb58158f429425fe950/torch/csrc/jit/ir/ir.cpp#L1936-L1938
> 
> In this example the IR after inlining would roughly look like:
> 
> ```
> graph(...):
>   %t1 = aten::add(...) # callstack: [A0::forward]
>   ...
>   %t2 = aten::add(%t1, ...) # callstack: [B0::forward]
>   ...
> ```
> 
> In order to get an example with a callstack of depth > 1, one would need to write something like this:
> 
> ```
>  class A(nn.Module):
>    def __init__(self):
>      super(A, self).__init__()
> 
>    def forward(self, x):
>      return x + 1
> 
>  class B(nn.Module):
>    def __init__(self):
>      super(B, self).__init__()
>      self.A_from_B = A()
> 
>    def forward(self, x):
>      return self.A_from_B.forward(x) + 2
> 
>  class C(nn.Module):
>    def __init__(self):
>      super(C, self).__init__()
>      self.B_from_C = B()
> 
>    def forward(self, x):
>      return self.B_from_C.forward(x)
> ```

@ZolotukhinM Thanks a lot for your explanation! I learned that https://github.com/pytorch/pytorch/blob/a03f05f2a28ee9516a26bab8e41351aeeb3a4d76/torch/csrc/jit/ir/scope.cpp#L108 is used to insert the link list! If `raw_callstack_ptr == NULL`, it will be the starting node, which means the top of the callstack. That answered my first confusion on the NULL pointer. How about the second one "2) It first inlines B.forward. When it goes A.forward, in addition to checking new_node = add node from A, it also checks new_node = inserted add node from B, which does not look proper to me."? For this one @cccclai verified that using the actual new node, instead of value_map, we could get the expected results. Is it likely a bug as @kimishpatel also asked?

Reply:
@kimishpatel this part is correct. When setting the value_map (`value_map[node->outputs()[i]] = new_node->outputs()[i];`) and node_map (`node_map[node] = new_node`), `new_node->outputs()[i]->node() == new_node` is true. However, when trying to access the value_map ([code pointer](https://github.com/pytorch/pytorch/pull/47416/commits/ce308cde64ce648984189610060ad828c18b63a7#diff-12d743f00439e8ccdc9575e788d00b60081dfbc8ce0a9e7b5184cb0d87843c2aR1954)) (vs node_map), it's different.

Reply:
To simplify the problem, should I expect to get the same node vector for these two different methods?
```
vec<Node*> nodes;
for (const auto& kv : value_map) {
    Node* new_node = kv.second->node();
    nodes.append(new_node);
}
```
```
vec<Node*> nodes;
for (const auto& kv : node_map) {
    Node* new_node = kv.second;
    nodes.append(new_node);
}
```

Reply:
I would say yes. It seems to be a problem that we are not seeing that.

Reply:
OK I debugged the issue. Real problem is a kind of bug in callstack generation.
The issue with using value_map is that value map contains values not all for the nodes of the graph but primary inputs as well:
So 
graph A(1, 2):
node1
node2
The value map of this will contain (1, 2, node1, node2) wheras node map will contain only node1, node2.

Now for us we have
B(x, y):
node1
node2
That gets inlined first. 
So we get new value x that is inline of B. This x in the input for A
like
x = B...
y = A(x)

And if A is
A(1):
node1
node2
So during ilining A the value map we get is (x, node1, node2):
Now since x is a node that is already inlined its callstack contains "call to B". Then during inlining A when we process values map we process x. Since x already has a callstack we think we are calling x from within and thus append "callee" which is A to the callstack of x.

So in some sense Chen's fix is right. We should not process primary inputs for callstack processing. It should only contain internal nodes.

Need to think more if we can do this without adding node_map to function signature.

cc: @ZolotukhinM @iseeyuan @cccclai

Reply:
The `value_map` and `node_map` would produce similar results, but not necessarily equivalent. In theory we should have both maps, but in practice we decided to have only `node_map` as it was sufficient for issues at hand - if now we need `value_map` as well it's ok to add it.

Using a `node_map` would differ from using `value_map` when the cloned graph contains a node with no inputs/outputs (this node would be accessible through the `node_map`, but not through the `value_map`) or when a value is not used in any of the nodes in the graph (the value will be accessible through the `value_map` but not through the `node_map`).

Reply:
> To simplify the problem, should I expect to get the same node vector for these two different methods?

It would differ in scenarios I mentioned above, also in the first loop some of the nodes could appear in the vector multiple times (if the have several outputs). The order in the vectors would also be different since the maps are unordered.

Reply:
> Need to think more if we can do this without adding node_map to function signature.
IIUC the issue here is that value maps from two inlines overlap - is it correct? If so, then an easy solution would be to skip values that are inputs when updating the callstacks since the overlap could only happen there (need to think carefully about how to properly handle these values, but it shouldn't be very hard I think).

Reply:
> In theory we should have both maps, but in practice we decided to have only `node_map` as it was sufficient for issues at hand - if now we need `value_map` as well it's ok to add it.

You mean the other way, to switch `node_map` and `value_map`, right?

Reply:
> IIUC the issue here is that value maps from two inlines overlap - is it correct? If so, then an easy solution would be to skip values that are inputs when updating the callstacks since the overlap could only happen there (need to think carefully about how to properly handle these values, but it shouldn't be very hard I think).

Right and overlap happens because output from the first inlined graph is input to the next one. So yes, if we skip inputs when updating callstacks that should be sufficient.

Reply:
> The value_map and node_map would produce similar results, but not necessarily equivalent. In theory we should have both maps, but in practice we decided to have only node_map as it was sufficient for issues at hand - if now we need value_map as well it's ok to add it.

You mean value_map was sufficient?

>Using a node_map would differ from using value_map when the cloned graph contains a node with no inputs/outputs (this node would be accessible through the node_map, but not through the value_map) or when a value is not used in any of the nodes in the graph (the value will be accessible through the value_map but not through the node_map).

Do we expect side effect methods without inputs/outputs? I presume to handle all scenarios then we should use both maps and figure out a way to inline without producing duplicates.

Reply:
> You mean the other way, to switch node_map and value_map, right?
> You mean value_map was sufficient?

Yes, whatever exists now :)

> Do we expect side effect methods without inputs/outputs?

I am thinking of ` = prim::Print()`, but that supposedly doesn't have side effects. From the top of my head I can't think of other examples (another thing I've just tried is `prim::RaiseException`, but it seems to have at least one input), but I'm not certain that they don't exist. I'd say if such cases exist and we don't update callstack info for them, it's not a very big deal :)

Reply:
I saw a TODO [(code)](https://github.com/pytorch/pytorch/pull/47416/files#diff-12d743f00439e8ccdc9575e788d00b60081dfbc8ce0a9e7b5184cb0d87843c2aR1920-R1921) saying 

> We might need to use nodes_map instead of value_map. Otherwise, we are missing nodes without outputs (e.g. prim::Print).

Will it be a potential issue if we only use `value_map`?

Reply:
It should not be an issue if we also visit inputs, but it will be an issue if we only visit outputs (like in the existing code).

But if we take a step back - could you please explain one more time why we need node_map instead of value_map? The differences I'm aware of shouldn't matter in the examples at hand, so I'm a bit confused. (This thread is super long and I kinda lost trying to find the answer there :) )

Reply:
Let me rephrase a little:
1) I think you've found a bug in the current implementation when we produce an incorrect call-stack for the cases where several calls were inlined and the corresponding value-maps overlapped.
2) Using node_map instead of value_map would help to address the TODO - i.e. we would be able to populate callstacks for nodes without outputs as well.
3) Augmenting callstacks with module info would give a better info to the user - that's what you originally tried to do :)

(Please correct me if my understanding is inaccurate.)

I think these three items are separate and while it'd be good to resolve all of them, you don't have to since (1) and (2) have existed before :) If you decide to fix all of them, that'd be awesome - but in this case I'd suggest splitting the fixes into separate PRs anyway.

Reply:
The original purpose of this pr is to get the module path by revisiting inlinedCallStack for each node after the graph is inlined [(the main part)](https://github.com/pytorch/pytorch/pull/47416/files#diff-00039b0bf9e101a8b2a13528c1748e1a5b75c5b47cdbab0d51a92b536a12c941R60-R89). However, the module path of this unit test seems not correct, and the reason is that `value_map` includes overlapping nodes. 
```
c10::make_intrusive<InlinedCallStack>(
                *new_node_cs,
                callee,
                to_replace->sourceRange(),
                module_instance_info);
)
```

From my understanding, `value_map` is used to iterate each node and set `inlinedCallstack` accordingly when inlining graph [(code).](https://github.com/pytorch/pytorch/pull/47416/files#diff-12d743f00439e8ccdc9575e788d00b60081dfbc8ce0a9e7b5184cb0d87843c2aR1923-R1952) If `value_map` includes extra nodes, these nodes will have wrong `inlinedCallstack`, which I think need to be fixed.

As a note: In my previous pr, I update inlinedCallstack to include module information. https://github.com/pytorch/pytorch/pull/46729

Reply:
> The original purpose of this pr is to get the module path by revisiting inlinedCallStack for each node after the graph is inlined (the main part). However, the module path of this unit test seems not correct, and the reason is that value_map includes overlapping nodes.

Yeah, what I'm saying is that fixing it could be factored out from the scope of this PR - just change the test to not expose that bug.

In a follow-up PR where you fix the issue with overlapping values - either by excluding them or by switching to `node_map` - you could add the test in the current form.

Reply:
Ah I see. Thank you that's clear. Looks like the browser doesn't re-render this page and I didn't see the rephrase comment above when typing my comment. :)

Reply:
@ZolotukhinM @cccclai I think doing this in separate PR is fine but lets stack on top of this. Part of the project is to get accurate module info and we will need to land both PRs.

With respect to how:
Probably using node_map is cleaner. Alternatively we can do this:
Since the issue is really coming from `Value`s which are input of the callee graph we can just inspect that while iterating over value_map in `inlineCallTo` as follows:
```
bool isValueGraphInput(const Value* key, const Graph& g) {
  for (const Value* v : g.inputs()) {
    if (key == v) {
      return true;
    }
  }
 return false;
}
for (const auto& kv : value_map) {
  if (isValueGraphInput(kv.first, callee.graph()) { /* or callee.optimized_graph()? Not sure if it makes a difference for checking inputs */
    continue;
  }
}
```

Reply:
I understand the logic of this piece of code and tried it. It seems not working out of box and I'm still investigating the reason. it looks like no value key matches the `callee.graph().inputs()` somehow.

Reply:
uh. Strange. I am not married to this method so its ok if you want to go with your approach. Just thought this might be little less intrusive.

Reply:
@kimishpatel I figure out the reason. I should use optimize_graph() instead of graph(), because optimzied_graph() is the input when setting value_map. The latest commit should work now. Feel free to let me know your thought.